processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -624,46 +626,67 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
     if (row.getTimestampFromEpoch() < minTimestamp) {
       throw new IAE("Cannot add row[%s] because it is below the minTimestamp[%s]", row, DateTimes.utc(minTimestamp));
     }
-    final List<String> rowDimensions = row.getDimensions();
+    DimensionData prevDimensionData = this.dimensions.get();
+    RowDimsKeyComponents dimsKeyComponents = getRowDimsKeyComponents(row, prevDimensionData);
+    DimensionData dimensionData = dimsKeyComponents.getUpdatedDimensionData();
+    while (dimensionData != null && !dimensions.compareAndSet(prevDimensionData, dimensionData)) {
```

Comment:
It's not documented what does it mean for `dimensionData` to be null here. Please add a Javadoc comment to `RowDimsKeyComponents`'s field.